lib/strong_migrations/migration_helpers.rb

```diff
@@ -73,16 +75,172 @@ def remove_null_constraint_safely(table_name, column_name, name: nil)
       end
     end
 
+    def add_column_safely(table_name, column_name, type, **options)
+      ensure_postgresql(__method__)
+      ensure_not_in_transaction(__method__)
+
+      default = options.delete(:default)
+
+      if postgresql_version >= 110000 || default.nil?
+        add_column(table_name, column_name, options)
+      else
+        reversible do |dir|
+          dir.up do
+            transaction do
+              add_column(table_name, column_name, type, default: nil, **options)
+              change_column_default(table_name, column_name, default)
+            end
+
+            default_after_type_cast = connection.type_cast(default)
+            backfill_column_safely(table_name, column_name, default_after_type_cast)
+
+            allow_null = options[:null]
+            add_null_constraint_safely(table_name, column_name) unless allow_null
+          end
+
+          dir.down do
+            remove_column(table_name, column_name)
+          end
+        end
+      end
+    end
+
+    def backfill_column_safely(table_name, column_name, value, batch_size: 1000)
+      ensure_not_in_transaction(__method__)
+
+      table = Arel::Table.new(table_name)
+      count_arel = table.project(Arel.star.count.as("count"))
+      total = connection.exec_query(count_arel.to_sql).first["count"]
+
+      return if total == 0
+
+      primary_key = connection.primary_key(table_name)
+
+      start_arel = table
+        .project(table[primary_key])
+        .order(table[primary_key].asc)
+        .take(1)
+
+      start_pk = connection.exec_query(start_arel.to_sql).first[primary_key]
+
+      loop do
+        finish_arel = table
+          .project(table[primary_key])
+          .where(table[primary_key].gteq(start_pk))
+          .order(table[primary_key].asc)
+          .skip(batch_size)
+          .take(1)
+
+        finish_result = connection.exec_query(finish_arel.to_sql).first
+
+        update_arel = Arel::UpdateManager.new
```

Comment:
Our app sometimes updates multiple rows which makes migrations like this deadlock prone. We always force locks to be acquired in primary key order to avoid this. Not sure how common that practice is and whether or not it's worth folding into the gem to avoid deadlocks.

Reply:
Didn't get it. Can you provide examples and elaborate a little more?